llvm/tools/llvm-foreach/llvm-foreach.cpp

```diff
@@ -0,0 +1,207 @@
+             ".out"),
+    cl::init("out"), cl::value_desc("R")};
+
+// Emit list of produced files for better integration with other tools.
+static cl::opt<std::string> OutputFileList{
+    "out-file-list", cl::desc("Specify filename for list of outputs."),
+    cl::value_desc("filename"), cl::init("")};
+
+static void error(const Twine &Msg) {
+  errs() << "llvm-foreach: " << Msg << '\n';
+  exit(1);
+}
+
+static void error(std::error_code EC, const Twine &Prefix) {
+  if (EC)
+    error(Prefix + ": " + EC.message());
+}
+
+int main(int argc, char **argv) {
+  cl::ParseCommandLineOptions(
+      argc, argv,
+      "llvm-foreach: Execute specified command as many times as\n"
+      "necessary to use up the list of input items.\n"
+      "Usage:\n"
+      "llvm-foreach --in-file-list=a.list --in-replace='{}' -- echo '{}'\n"
+      "NOTE: commands containig redirects are not supported by llvm-foreach\n"
+      "yet.\n");
+
+  ExitOnError ExitOnErr("llvm-foreach: ");
+
+  if (InputFileLists.size() != Replaces.size())
+    error("Number of input file lists and input path replaces don't match.");
+
+  std::vector<std::unique_ptr<MemoryBuffer>> MBs;
+  std::vector<line_iterator> LineIterators;
+  for (auto &InputFileList : InputFileLists) {
+    std::unique_ptr<MemoryBuffer> MB = ExitOnErr(
+        errorOrToExpected(MemoryBuffer::getFileOrSTDIN(InputFileList)));
+    LineIterators.push_back(line_iterator(*MB));
+    MBs.push_back(std::move(MB));
+  }
+
+  SmallVector<StringRef, 8> Args(InputCommandArgs.begin(),
+                                 InputCommandArgs.end());
+
+  if (Args.empty())
+    error("No command?");
+
+  struct ArgumentReplace {
+    int ArgNum = -1;
+    std::string Prefix;
+    std::string Postfix;
+  };
+  auto CreateArgumentReplace = [](int ArgNum, StringRef Arg,
+                                  StringRef Replace) -> ArgumentReplace {
+    std::string Prefix = Arg.substr(/*Start*/ 0, Arg.find(Replace));
+    std::string Postfix = Arg.substr(Arg.find_last_of(Replace) + 1);
+    return {ArgNum, Prefix, Postfix};
+  };
+
+  // Find args to replace with filenames from input list.
+  std::vector<ArgumentReplace> InReplaceArgs;
+  ArgumentReplace OutReplaceArg;
+  for (size_t i = 0; i < Args.size(); ++i) {
+    for (auto &Replace : Replaces)
+      if (Args[i].contains(Replace)) {
+        InReplaceArgs.push_back(CreateArgumentReplace(i, Args[i], Replace));
+      }
+
+    if (!OutReplace.empty() && Args[i].contains(OutReplace))
+      OutReplaceArg = CreateArgumentReplace(i, Args[i], OutReplace);
+  }
+
+  // Emit an error if user requested replace output file in the command but
+  // replace string is not found.
+  if (!OutReplace.empty() && OutReplaceArg.ArgNum < 0)
+    error("Couldn't find replace string for output in the command.");
+
+  // Make sure that specified program exists, emit an error if not.
+  std::string Prog =
+      ExitOnErr(errorOrToExpected(sys::findProgramByName(Args[0])));
```

Comment:
What happens with `sys::ExecuteAndWait` if we don't do that?

Reply:
An error will be emitted.
Note: This also helps to simplify usage of tool, because without it `ExecuteAndWait` will fail if just only name (instead of full path) of executable is provided.
i.e.:
`llvm-foreach ... - echo something`
will fail without `findProgramByName` usage, and something like this only will work:
`llvm-foreach ... - /full/path/to/echo something`